File: android/app/src/main/kotlin/com/bansheerun/ActivityRepository.kt
package com.bansheerun

import android.content.Context
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

/**
 * Repository for storing and loading activities and personal bests.
 */
class ActivityRepository(private val context: Context) {

    private val json = Json { ignoreUnknownKeys = true }

    private val baseDir: File
        get() {
            val dir = File(context.filesDir, "banshee_run")
            if (!dir.exists()) dir.mkdirs()
            return dir
        }

    private val activitiesDir: File
        get() {
            val dir = File(baseDir, "activities")
            if (!dir.exists()) dir.mkdirs()
            return dir
        }

    private val indexFile: File
        get() = File(activitiesDir, "index.json")

    private val personalBestsFile: File
        get() = File(baseDir, "personal_bests.json")

    // ============================================================================
    // Activity Index
    // ============================================================================

    fun loadActivityIndex(): ActivityIndex {
        return try {
            if (indexFile.exists()) {
                json.decodeFromString(indexFile.readText())
            } else {
                ActivityIndex(emptyList())
            }
        } catch (_: Exception) {
            ActivityIndex(emptyList())
        }
    }

    private fun saveActivityIndex(index: ActivityIndex) {
        indexFile.writeText(json.encodeToString(index))
    }

    // ============================================================================
    // Activities
    // ============================================================================

    fun saveActivity(activityJson: String) {
        // Parse activity to get ID
        val activity = try {
            json.decodeFromString<Activity>(activityJson)
        } catch (_: Exception) {
            return
        }

        // Save full activity to file
        val activityFile = File(activitiesDir, "${activity.id}.json")
        activityFile.writeText(activityJson)

        // Get summary and add to index
        val summaryJson = BansheeLib.getActivitySummary(activityJson) ?: return
        val summary = try {
            json.decodeFromString<ActivitySummary>(summaryJson)
        } catch (_: Exception) {
            return
        }

        // Update index
        val index = loadActivityIndex()
        val newActivities = listOf(summary) + index.activities
        saveActivityIndex(ActivityIndex(newActivities))

        // Update PBs
        val existingPbs = loadPersonalBestsJson()
        val updatedPbs = BansheeLib.updatePbs(existingPbs, activityJson)
        if (updatedPbs != null) {
            savePersonalBestsJson(updatedPbs)
        }
    }

    fun loadActivity(id: String): String? {
        val activityFile = File(activitiesDir, "$id.json")
        return if (activityFile.exists()) activityFile.readText() else null
    }

    fun deleteActivity(id: String) {
        // Remove from index
        val index = loadActivityIndex()
        val newActivities = index.activities.filter { it.id != id }
        saveActivityIndex(ActivityIndex(newActivities))

        // Delete file
        val activityFile = File(activitiesDir, "$id.json")
        if (activityFile.exists()) activityFile.delete()
    }

    // ============================================================================
    // Personal Bests
    // ============================================================================

    fun loadPersonalBestsJson(): String? {
        return if (personalBestsFile.exists()) {
            personalBestsFile.readText()
        } else {
            null
        }
    }

    private fun savePersonalBestsJson(pbsJson: String) {
        personalBestsFile.writeText(pbsJson)
    }

    fun loadPersonalBests(): List<PersonalBest> {
        val pbsJson = loadPersonalBestsJson() ?: return emptyList()
        return try {
            val pbs = json.decodeFromString<PersonalBestsWrapper>(pbsJson)
            pbs.records
        } catch (_: Exception) {
            emptyList()
        }
    }

    fun getPersonalBestsForType(activityType: BansheeLib.ActivityType): List<PersonalBest> {
        val pbsJson = loadPersonalBestsJson() ?: return emptyList()
        val filteredJson = BansheeLib.getPbsForType(pbsJson, activityType.value) ?: return emptyList()
        return try {
            json.decodeFromString(filteredJson)
        } catch (_: Exception) {
            emptyList()
        }
    }

    fun getNewPBs(activityJson: String): List<PersonalBest> {
        val existingPbs = loadPersonalBestsJson()
        val newPbsJson = BansheeLib.getNewPbs(existingPbs, activityJson) ?: return emptyList()
        return try {
            json.decodeFromString(newPbsJson)
        } catch (_: Exception) {
            emptyList()
        }
    }

    // ============================================================================
    // Filtering
    // ============================================================================

    fun getActivities(activityType: BansheeLib.ActivityType? = null): List<ActivitySummary> {
        val index = loadActivityIndex()
        return if (activityType != null) {
            index.activities.filter { it.activityType == activityType.name.lowercase() }
        } else {
            index.activities
        }
    }

    companion object {
        @Volatile
        private var instance: ActivityRepository? = null

        fun getInstance(context: Context): ActivityRepository {
            return instance ?: synchronized(this) {
                instance ?: ActivityRepository(context.applicationContext).also {
                    instance = it
                }
            }
        }
    }
}

// ============================================================================
// Data Classes
// ============================================================================

@Serializable
data class ActivityIndex(
    val activities: List<ActivitySummary>
)

@Serializable
data class ActivitySummary(
    val id: String,
    val name: String,
    val activity_type: String,
    val total_distance_meters: Double,
    val duration_ms: Long,
    val recorded_at: Long,
    val pace_min_per_km: Double
) {
    val activityType: String get() = activity_type
    val totalDistanceMeters: Double get() = total_distance_meters
    val durationMs: Long get() = duration_ms
    val recordedAt: Long get() = recorded_at
    val paceMinPerKm: Double get() = pace_min_per_km

    fun getActivityTypeEnum(): BansheeLib.ActivityType? {
        return BansheeLib.ActivityType.entries.find { it.name.lowercase() == activity_type }
    }
}

@Serializable
data class Activity(
    val id: String,
    val name: String,
    val activity_type: String,
    val coordinates: List<Coordinate>,
    val total_distance_meters: Double,
    val duration_ms: Long,
    val recorded_at: Long
)

@Serializable
data class Coordinate(
    val lat: Double,
    val lon: Double,
    val timestamp_ms: Long
)

@Serializable
data class PersonalBestsWrapper(
    val records: List<PersonalBest>
)

@Serializable
data class PersonalBest(
    val activity_type: String,
    val distance_meters: Double,
    val time_ms: Long,
    val activity_id: String,
    val achieved_at: Long,
    val pace_min_per_km: Double
) {
    val activityType: String get() = activity_type
    val distanceMeters: Double get() = distance_meters
    val timeMs: Long get() = time_ms
    val activityId: String get() = activity_id
    val achievedAt: Long get() = achieved_at
    val paceMinPerKm: Double get() = pace_min_per_km

    fun getDistanceName(): String = BansheeLib.getDistanceName(distance_meters) ?: "Custom"
    fun getFormattedTime(): String = BansheeLib.formatDuration(time_ms) ?: "0:00"
    fun getFormattedPace(): String = BansheeLib.formatPace(distance_meters, time_ms) ?: "0:00 /km"
}

File: macos/BansheeRun/ActivityRepository.swift
import Foundation

/// Repository for storing and loading activities and personal bests.
class ActivityRepository: ObservableObject {
    static let shared = ActivityRepository()

    @Published var activities: [ActivitySummary] = []
    @Published var personalBests: String = "{\"records\":[]}"

    private let fileManager = FileManager.default

    private var baseURL: URL {
        let appSupport = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let bansheeDir = appSupport.appendingPathComponent("BansheeRun", isDirectory: true)

        // Create directory if it doesn't exist
        if !fileManager.fileExists(atPath: bansheeDir.path) {
            try? fileManager.createDirectory(at: bansheeDir, withIntermediateDirectories: true)
        }

        return bansheeDir
    }

    private var activitiesDir: URL {
        let dir = baseURL.appendingPathComponent("activities", isDirectory: true)
        if !fileManager.fileExists(atPath: dir.path) {
            try? fileManager.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        return dir
    }

    private var indexURL: URL {
        activitiesDir.appendingPathComponent("index.json")
    }

    private var personalBestsURL: URL {
        baseURL.appendingPathComponent("personal_bests.json")
    }

    init() {
        loadAll()
    }

    // MARK: - Load Operations

    func loadAll() {
        loadActivityIndex()
        loadPersonalBests()
    }

    private func loadActivityIndex() {
        guard fileManager.fileExists(atPath: indexURL.path),
              let data = try? Data(contentsOf: indexURL),
              let _ = String(data: data, encoding: .utf8) else {
            activities = []
            return
        }

        // Parse JSON into ActivitySummary array
        if let decoded = try? JSONDecoder().decode(ActivityIndex.self, from: data) {
            activities = decoded.activities
        }
    }

    private func loadPersonalBests() {
        guard fileManager.fileExists(atPath: personalBestsURL.path),
              let data = try? Data(contentsOf: personalBestsURL),
              let json = String(data: data, encoding: .utf8) else {
            personalBests = "{\"records\":[]}"
            return
        }
        personalBests = json
    }

    // MARK: - Save Operations

    func saveActivity(activityJson: String) {
        guard let activityData = activityJson.data(using: .utf8),
              let activity = try? JSONDecoder().decode(Activity.self, from: activityData) else {
            return
        }

        // Save full activity to individual file
        let activityURL = activitiesDir.appendingPathComponent("\(activity.id).json")
        try? activityData.write(to: activityURL)

        // Get summary and add to index
        if let summaryJson = BansheeLib.getActivitySummary(activityJson: activityJson),
           let summaryData = summaryJson.data(using: .utf8),
           let summary = try? JSONDecoder().decode(ActivitySummary.self, from: summaryData) {
            activities.insert(summary, at: 0)
            saveActivityIndex()
        }

        // Update PBs
        if let updatedPbs = BansheeLib.updatePbs(existingPbsJson: personalBests, activityJson: activityJson) {
            personalBests = updatedPbs
            savePersonalBests()
        }
    }

    private func saveActivityIndex() {
        let index = ActivityIndex(activities: activities)
        if let data = try? JSONEncoder().encode(index) {
            try? data.write(to: indexURL)
        }
    }

    private func savePersonalBests() {
        if let data = personalBests.data(using: .utf8) {
            try? data.write(to: personalBestsURL)
        }
    }

    // MARK: - Load Single Activity

    func loadActivity(id: String) -> String? {
        let activityURL = activitiesDir.appendingPathComponent("\(id).json")
        guard let data = try? Data(contentsOf: activityURL),
              let json = String(data: data, encoding: .utf8) else {
            return nil
        }
        return json
    }

    // MARK: - Delete Activity

    func deleteActivity(id: String) {
        // Remove from index
        activities.removeAll { $0.id == id }
        saveActivityIndex()

        // Delete file
        let activityURL = activitiesDir.appendingPathComponent("\(id).json")
        try? fileManager.removeItem(at: activityURL)
    }

    // MARK: - Filter & Sort

    func getActivities(type: BansheeLib.ActivityType?) -> [ActivitySummary] {
        if let type = type {
            return activities.filter { $0.activityType == type.rawValue }
        }
        return activities
    }

    func getPersonalBests(type: BansheeLib.ActivityType) -> [PersonalBest] {
        guard let filtered = BansheeLib.getPbsForType(pbsJson: personalBests, activityType: type),
              let data = filtered.data(using: .utf8),
              let pbs = try? JSONDecoder().decode([PersonalBest].self, from: data) else {
            return []
        }
        return pbs
    }

    func getNewPBs(activityJson: String) -> [PersonalBest] {
        guard let newPbsJson = BansheeLib.getNewPbs(existingPbsJson: personalBests, activityJson: activityJson),
              let data = newPbsJson.data(using: .utf8),
              let pbs = try? JSONDecoder().decode([PersonalBest].self, from: data) else {
            return []
        }
        return pbs
    }
}

// MARK: - Codable Models

struct ActivityIndex: Codable {
    var activities: [ActivitySummary]
}

struct ActivitySummary: Codable, Identifiable {
    let id: String
    let name: String
    let activityType: Int
    let totalDistanceMeters: Double
    let durationMs: Int64
    let recordedAt: Int64
    let paceMinPerKm: Double

    enum CodingKeys: String, CodingKey {
        case id, name
        case activityType = "activity_type"
        case totalDistanceMeters = "total_distance_meters"
        case durationMs = "duration_ms"
        case recordedAt = "recorded_at"
        case paceMinPerKm = "pace_min_per_km"
    }

    var type: BansheeLib.ActivityType {
        BansheeLib.ActivityType(rawValue: Int32(activityType)) ?? .run
    }
}

struct Activity: Codable {
    let id: String
    let name: String
    let activityType: String
    let coordinates: [Coordinate]
    let totalDistanceMeters: Double
    let durationMs: Int64
    let recordedAt: Int64

    enum CodingKeys: String, CodingKey {
        case id, name, coordinates
        case activityType = "activity_type"
        case totalDistanceMeters = "total_distance_meters"
        case durationMs = "duration_ms"
        case recordedAt = "recorded_at"
    }
}

struct Coordinate: Codable {
    let lat: Double
    let lon: Double
    let timestampMs: Int64

    enum CodingKeys: String, CodingKey {
        case lat, lon
        case timestampMs = "timestamp_ms"
    }
}

struct PersonalBest: Codable, Identifiable {
    let activityType: String
    let distanceMeters: Double
    let timeMs: Int64
    let activityId: String
    let achievedAt: Int64
    let paceMinPerKm: Double

    var id: String { "\(activityType)_\(distanceMeters)" }

    enum CodingKeys: String, CodingKey {
        case activityType = "activity_type"
        case distanceMeters = "distance_meters"
        case timeMs = "time_ms"
        case activityId = "activity_id"
        case achievedAt = "achieved_at"
        case paceMinPerKm = "pace_min_per_km"
    }

    var distanceName: String {
        BansheeLib.getDistanceName(distanceMeters: distanceMeters)
    }

    var formattedTime: String {
        BansheeLib.formatDuration(durationMs: timeMs)
    }

    var formattedPace: String {
        BansheeLib.formatPace(distanceMeters: distanceMeters, durationMs: timeMs)
    }
}
